core/lib/src/request/request.rs

```diff
@@ -78,6 +81,39 @@ impl<'r> Request<'r> {
         f(&mut request);
     }
 
+    /// Retrieves the cached value for type `T` from the request-local cached
+    /// state of `self`. If no such value has previously been cached for
+    /// this request, `f` is called to produce the value which is subsequently
+    /// returned.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use rocket::http::Method;
+    /// # use rocket::Request;
+    /// # struct User();
+    /// fn current_user(_: &Request) -> User {
+    ///     // Load user...
+    ///     # User()
+    /// }
+    ///
+    /// # Request::example(Method::Get, "/uri", |request| {
+    /// let user = request.local_cache(current_user);
+    /// # });
+    /// ```
+    pub fn local_cache<T, F>(&self, f: F) -> &T
+        where T: Send + Sync + 'static,
+              F: FnOnce(&Request) -> T {
```

Comment:
On second thought: why are we passing `self` to `F`? The caller already has a handle to `request`, so they can use it if they need to. I think this should probably be `F: FnOnce() -> T`.